pkg/backup/scheduler.go

```diff
@@ -0,0 +1,193 @@
+	glog.Info("Starting Stash backup")
+
+	go c.rInformer.Run(stopCh)
+
+	// Wait for all involved caches to be synced, before processing items from the queue is started
+	if !cache.WaitForCacheSync(stopCh, c.rInformer.HasSynced) {
+		runtime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runResticWatcher, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Info("Stopping Stash backup")
+}
+
+func (c *Controller) configureScheduler(r *api.Restic) error {
+	// Remove previous jobs
+	for _, v := range c.cron.Entries() {
+		c.cron.Remove(v.ID)
+	}
+	_, err := c.cron.AddFunc(r.Spec.Schedule, func() {
+		if err := c.runOnceForScheduler(); err != nil {
+			c.recorder.Event(r.ObjectReference(), core.EventTypeWarning, eventer.EventReasonFailedCronJob, err.Error())
+			log.Errorln(err)
+		}
+	})
+	if err != nil {
+		return err
+	}
+	_, err = c.cron.AddFunc("0 0 */3 * *", func() { c.checkOnceForScheduler() })
+	return err
+}
+
+func (c *Controller) runOnceForScheduler() error {
+	select {
+	case <-c.locked:
+		log.Infof("Acquired lock for Restic %s/%s", c.opt.Namespace, c.opt.ResticName)
+		defer func() {
+			c.locked <- struct{}{}
+		}()
+	default:
+		log.Warningf("Skipping backup schedule for Restic %s/%s", c.opt.Namespace, c.opt.ResticName)
+		return nil
+	}
+
+	// check resource again, previously done in setup()
+	resource, err := c.rLister.Restics(c.opt.Namespace).Get(c.opt.ResticName)
+	if kerr.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	if resource.Spec.Backend.StorageSecretName == "" {
+		return errors.New("missing repository secret name")
+	}
+	secret, err := c.k8sClient.CoreV1().Secrets(resource.Namespace).Get(resource.Spec.Backend.StorageSecretName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// setup restic again, previously done in setup()
+	if err = c.resticCLI.SetupEnv(resource, secret, c.opt.SmartPrefix); err != nil {
+		return err
+	}
+	if err = c.resticCLI.InitRepositoryIfAbsent(); err != nil {
+		return err
+	}
+
+	// run final restic backup command
+	return c.runResticBackup(resource)
+}
+
+func (c *Controller) checkOnceForScheduler() (err error) {
+	select {
+	case <-c.locked:
+		log.Infof("Acquired lock for Restic %s/%s", c.opt.Namespace, c.opt.ResticName)
+		defer func() {
+			c.locked <- struct{}{}
+		}()
+	default:
+		log.Warningf("Skipping checkup schedule for Restic %s/%s", c.opt.Namespace, c.opt.ResticName)
+		return
+	}
+
+	var resource *api.Restic
+	resource, err = c.rLister.Restics(c.opt.Namespace).Get(c.opt.ResticName)
+	if kerr.IsNotFound(err) {
+		err = nil
+		return
+	} else if err != nil {
+		return
+	}
+
+	err = c.resticCLI.Check()
+	if err != nil {
+		c.recorder.Eventf(resource.ObjectReference(), core.EventTypeWarning, eventer.EventReasonFailedToCheck, "Repository check failed for workload %s %s/%s. Reason: %v", c.opt.Workload.Kind, c.opt.Namespace, c.opt.Workload.Name, err)
```

Comment:
When the line is this long, move each parameter to a separate line. Makes it more readable.